### PR TITLE
For NDK/r22, correct LD environment variable.

### LINKED ISSUE #7953: [package] android-ndk/r22: Bug after update android recipe
### Package and Environment Details (include every applicable attribute)
  * Package Name/Version: **android-ndk/r22**
  * Operating System+version: **Linux Ubuntu 18.04**
  * Compiler+version: **GCC 8**
  * Docker image: **conanio/gcc8**
  * Conan version: **conan latest**
  * Python version: **Python 3.7.4**

### Conan profile (output of `conan profile show default` or `conan profile show <profile>` if custom profile is in use)
```
include(default)

[settings]
os=Android
os.api_level=21
compiler=clang
compiler.version=11
compiler.libcxx=c++_static
arch=armv7

[build_requires]
*: android-ndk/r22

```

### Logs (Include/Attach if Applicable)
<details><summary>Click to expand log</summary>

There is a problem on Android NDK R22 recipe after update to latest commit:

```
ERROR: android-ndk/r22: Error in package_info() method, line 266
	self.env_info.LD = self._define_tool_var("LD", "ld")
while calling '_define_tool_var', line 168
	raise ConanException(f"'Environment variable {name} could not be created: '{path}'")
	ConanException: 'Environment variable LD could not be created: '/home/runner/.conan/data/android-ndk/r22/_/_/package/4db1be536558d833e52e862fd84d64d75c2b3656/toolchains/llvm/prebuilt/linux-x86_64/bin/arm-linux-androideabi-ld'
COMMAND: conan install /home/runner/work/ezored/ezored/files/targets/android/conan/recipe/conanfile.py -pr:b default -pr:h ezored_android_profile -s:h arch=armv7 -s:h os.api_level=16 -s:h build_type=Debug -o ezored_arch=armv7 -o ezored_name=ezored -o ezored_version=1.0.0 --build=missing --update
WORKING DIR: /home/runner/work/ezored/ezored/build/android/Debug/armv7/conan
```

</details>

FULL LOG:
https://github.com/ezored/ezored/runs/4108829881?check_suite_focus=true#step:12:182

### PR DESCRIPTION
**Pull request by andioz**
Specify library name and version:  **android_ndk/r22**

Try to fix #7953 by relaxing the checks, but only for the `LD` variable. As far as I have seen, there is no direct executable in the previous recipe revision at this location for r22 too. I guess the correct tool is located at runtime with the additonal option `-fuse-ld=gold` respectively with an appropriate default.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by andioz** [comment]
This is an interesting case. Building this package work of course, but this is typically a build dependency, and in case before this fix it failed at runtime. Are there chances to add tests for **using** the build dependency in the CI runner pipeline?

File: recipes/android-ndk/all/conanfile.py [hunk 1]
```diff
@@ -263,7 +263,7 @@ def package_info(self):
             self.env_info.READELF = self._define_tool_var("READELF", "readelf", True)
             # there doesn't seem to be an 'elfedit' included anymore.
         else:
-            self.env_info.LD = self._define_tool_var("LD", "ld")
+            self.env_info.LD = self._define_tool_var("LD", "ld", skip_check=True)
```
Comment (madebr) [inline recipes/android-ndk/all/conanfile.py:266]: Disabling a failing check will result in a bugged package because `LD` will be incorrect.
In https://github.com/conan-io/conan-center-index/issues/7953#issuecomment-962421225, you wrote that `ld` is missing, but there exists a bfd and gold linker.

So I propose the following fix:
add a new method (or modify the existing one) to check for different suffices.
e.g.
```python
self.env_info.LD = self._define_tool_var("LD", "ld", possible_suffixes=["", ".bfd", ".gold"])
```
`_define_tool_var` should then try all sufixes and use the first successful one.

Comment (andioz) [reply recipes/android-ndk/all/conanfile.py:266]: I see your point, but I'm not sure if this will result in the correct setting. My aim here was to "restore" the old behavior for a recently introduced check (which is not working here). I don't know much about Android NDK, and I hope the original author of this part of code knows more about. 

My way of thinking was: `LD` is not a path to a tool, it is only the " path stem", which will be extended by the caller with the appropriate suffix. But after reviewing this idea multiple times I'm less convinced ;)

I played around a little bit with the `gtest` example, and finally I recognized that none of this env variables is used. Because of CMake I'm sure. So this are settings for other, low-level build scripts. 

I only found one reference inside the NDK package:



```makefile
# build/tools/common-build-host-funcs.sh line 697ff

bh_setup_host_env ()
{
    CC=$BH_HOST_CONFIG-gcc
    CXX=$BH_HOST_CONFIG-g++
    LD=$BH_HOST_CONFIG-ld
    AR=$BH_HOST_CONFIG-ar
    AS=$BH_HOST_CONFIG-as
    RANLIB=$BH_HOST_CONFIG-ranlib
    NM=$BH_HOST_CONFIG-nm
    STRIP=$BH_HOST_CONFIG-strip
    STRINGS=$BH_HOST_CONFIG-strings
    export CC CXX LD AR AS RANLIB NM STRIP STRINGS
...
```

Maybe this is the source for this part of code.

OK, in short: I can do that, but I'm not sure if it is the right thing, and which one of the two (bld or gold) should be taken? And for other platforms, are there other suffixes/defaults?

Maybe someone with more knowledge of NDK internals can give a clue.

Comment (madebr) [reply recipes/android-ndk/all/conanfile.py:266]: True, it is mostly autotools-type + custom build scripts that are of concern here, CMake uses CXX/CC as the linker.

You detected a real bug here: an environment variable with an invalid value.
So I feel uncomfortable ignoring it.

On the other hand, I think it would be better to just not define `LD`.
As I've written above, most build systems use `CC/CXX` as default linker executable.

This would also avoid a lot of warnings/errors causes by build scripts passing compiler flags to the linker.
e.g. binutil's linkers doesn't understand `-D` macro definitions (they don't make sense when linking)

So please just remove the `LD` definitions, and add a comment instead.
Something like:
```
# don't define LD to $prefix-ld/$prefix-ld.gold/$prefix-bfd. Build scripts should use the compiler (CC/CXX) as frontend.
```

File: recipes/android-ndk/all/conanfile.py [hunk 2]
```diff
@@ -263,7 +263,9 @@ def package_info(self):
             self.env_info.READELF = self._define_tool_var("READELF", "readelf", True)
             # there doesn't seem to be an 'elfedit' included anymore.
         else:
-            self.env_info.LD = self._define_tool_var("LD", "ld")
+            # don't define LD to $prefix-ld/$prefix-ld.gold/$prefix-bfd.
+            # Build scripts should use the compiler (CC/CXX) as frontend.
+            # self.env_info.LD = self._define_tool_var("LD", "ld")
```
Comment (madebr) [inline recipes/android-ndk/all/conanfile.py:268]:
```suggestion
            # Don't define LD to $prefix-ld/$prefix-ld.gold/$prefix-ld.bfd:
            # build scripts should use the compiler (CC/CXX) as frontend.
            # self.env_info.LD = self._define_tool_var("LD", "ld")
```

**Comment by justinmcbride** [comment]
See my comment in the issue that this ticket is tracked from.

It seems that the `ld` environment variable had been always missing in NDK 22b (it changed naming convention from 21e -> 22b), but adding in the exception made it clear that was the case. It has the unfortunate side effect of breaking "working" installations. 

Here is a list of the `ld` tools from my local digging:
```
./android-ndk-r21e-darwin-x86_64/toolchains/llvm/prebuilt/darwin-x86_64/bin/aarch64-linux-android-ld
./android-ndk-r21e-darwin-x86_64/toolchains/llvm/prebuilt/darwin-x86_64/bin/arm-linux-androideabi-ld
./android-ndk-r21e-darwin-x86_64/toolchains/llvm/prebuilt/darwin-x86_64/bin/i686-linux-android-ld
./android-ndk-r21e-darwin-x86_64/toolchains/llvm/prebuilt/darwin-x86_64/bin/x86_64-linux-android-ld
./android-ndk-r21e-linux-x86_64/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android-ld
./android-ndk-r21e-linux-x86_64/toolchains/llvm/prebuilt/linux-x86_64/bin/arm-linux-androideabi-ld
./android-ndk-r21e-linux-x86_64/toolchains/llvm/prebuilt/linux-x86_64/bin/i686-linux-android-ld
./android-ndk-r21e-linux-x86_64/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android-ld
./android-ndk-r21e-windows-x86_64/toolchains/llvm/prebuilt/windows-x86_64/bin/aarch64-linux-android-ld.exe
./android-ndk-r21e-windows-x86_64/toolchains/llvm/prebuilt/windows-x86_64/bin/arm-linux-androideabi-ld.exe
./android-ndk-r21e-windows-x86_64/toolchains/llvm/prebuilt/windows-x86_64/bin/i686-linux-android-ld.exe
./android-ndk-r21e-windows-x86_64/toolchains/llvm/prebuilt/windows-x86_64/bin/x86_64-linux-android-ld.exe
./android-ndk-r22b-darwin-x86_64/toolchains/llvm/prebuilt/darwin-x86_64/bin/ld
./android-ndk-r22b-linux-x86_64/toolchains/llvm/prebuilt/linux-x86_64/bin/ld
./android-ndk-r22b-windows-x86_64/toolchains/llvm/prebuilt/windows-x86_64/bin/ld.exe
./android-ndk-r23b-darwin/toolchains/llvm/prebuilt/darwin-x86_64/bin/ld
./android-ndk-r23b-linux/toolchains/llvm/prebuilt/linux-x86_64/bin/ld
./android-ndk-r23b-windows/toolchains/llvm/prebuilt/windows-x86_64/bin/ld.exe
```

I'd suggest either simply removing the exception from being raised (because I'm not sure what other tools changed) or changing the `LD` line to check for version 22 and leaving the others checked for r23.
```
if tools.Version(self._ndk_version) >= 22:
    self.env_info.LD = self._define_tool_var_naked("LD", "ld")
````

**Comment by andioz** [comment]
Thanks @justinmcbride for the details! As an addition, I add this link to the NDK docs, it shows an example for using NDK (Version r22 and above!) with Autoconv: https://developer.android.com/ndk/guides/other_build_systems#autoconf

In this example `LD` gets no `$TARGET$API` prfefix like other tools. I will commit a change which uses @justinmcbride's suggestion for changing only for r22+ (r23 and above is still unchanged). This works for me on Linux with r22.

I think this will be better supporting existing r21 projects.

**Review by justinmcbride — APPROVED** [review]
To be honest, I don't love the control flow there, but that's nit-picking something that probably works and doesn't need to be over-optimized. Also, I just noticed last night while doing this investigating that this recipe only provides r22, not the newer patch r22b. So I have plans to add support for that this week (for my own use as well), and may look into adding proper testing and whatnot along the way. Thanks for fixing the error that my PR introduced.

**Review by SSE4 — APPROVED** [review]

**Comment by andioz** [comment]
Yeah it’s not really pretty code. But without a deep knowledge of all possible combinations I think it’s better to touch only the known issue at the moment. Oops, I missed that issue with version number comparison too, thought implicitly it’s a string compare. 

Would be great if you can spend time to clean it up later!

**Comment by paulo-coutinho** [comment]
I think that you need close/open the issue to solve the unexpected error.

**Comment by paulo-coutinho** [comment]
Very complicated. This is very buggy :(

**Comment by andioz** [comment]
I had to add a dummy commit to re-trigger CI. Sorry for the inconvenience...

**Comment by andioz** [comment]
No chance, need advice. Continue with this one or shall I discard and start a new l one?

**Comment by paulo-coutinho** [comment]
Hi @uilianries, can you help us understand if the problem is on CI?

**Comment by SSE4** [comment]
> Hi @uilianries, can you help us understand if the problem is on CI?

unfortunetely, there is a random bug in our CI scripts, it has nothing to do with this PR or recipe. it's generic problem with CI state serialization, we're going to deploy hotfix soon. sorry for the inconvinience. I'll restart builds once we have a fix.

**Review by SSE4 — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
LGTM

**Comment by andioz** [comment]
@madebr, @justinmcbride we would need a third approval for finishing the task

**Review by madebr — APPROVED** [review]

**Comment by ShuangLiu1992** [comment]
Still hitting the same error on Windows (not an issue on Macos)

**Comment by martin-schulze-vireso** [comment]
I think I have hit the same issue as @ShuangLiu1992 on Linux. I tried to dig further into it and noticed that the r22 branch of this fix is not taken on my system. I did the following test with the SemVer library that is used by conans version tools (https://github.com/conan-io/conan/blob/f230bf104842c160be6ea377497151fe1ea64c38/conans/client/tools/version.py#L48):

```
>>> SemVer("22b", True).compare("22")
-1 # this is the problem, SemVer thinks 22b < 22
# some more outputs for comparison
>>> SemVer("22b", True).compare("21")
1
>>> SemVer("22b", True).compare("22")
-1
>>> SemVer("22b", True).compare("22a")
1
```
The code expects that `Version("22b") >= 22` but SemVer/Conan thinks that `Version("22b") < 22`, thus not triggering the correct path.
